fe/fe-core/src/main/java/com/starrocks/qe/scheduler/Deployer.java

```diff
@@ -0,0 +1,207 @@
+
+    public Deployer(ConnectContext context,
+                    JobSpec jobSpec,
+                    ExecutionDAG executionDAG,
+                    TNetworkAddress coordAddress,
+                    FailureHandler failureHandler) {
+        this.jobSpec = jobSpec;
+        this.executionDAG = executionDAG;
+
+        this.tFragmentInstanceFactory = new TFragmentInstanceFactory(context, jobSpec, executionDAG, coordAddress);
+        this.emptyDescTable = new TDescriptorTable()
+                .setIs_cached(true)
+                .setTupleDescriptors(Collections.emptyList());
+
+        TQueryOptions queryOptions = jobSpec.getQueryOptions();
+        this.deliveryTimeoutMs = Math.min(queryOptions.query_timeout, queryOptions.query_delivery_timeout) * 1000L;
+
+        this.failureHandler = failureHandler;
+    }
+
+    public void deployFragments(List<ExecutionFragment> concurrentFragments) throws RpcException, UserException {
+        // Divide requests of fragments in the current group to two stages.
+        // - stage 1, the first request to a host, which need send descTable.
+        // - stage 2, the non-first requests to a host, which needn't send descTable.
+        List<List<FragmentInstanceExecState>> twoStageExecutionsToDeploy = ImmutableList.of(new ArrayList<>(), new ArrayList<>());
+
+        concurrentFragments.forEach(fragment -> this.createFragmentInstanceExecStates(fragment, twoStageExecutionsToDeploy));
+
+        for (List<FragmentInstanceExecState> executions : twoStageExecutionsToDeploy) {
+            executions.forEach(FragmentInstanceExecState::deployAsync);
+
+            waitForDeploymentCompletion(executions);
+        }
+    }
+
+    public interface FailureHandler {
+        void apply(Status status, FragmentInstanceExecState execution, Throwable failure) throws RpcException, UserException;
+    }
+
+    private void createFragmentInstanceExecStates(ExecutionFragment fragment,
+                                                  List<List<FragmentInstanceExecState>> twoStageExecutionsToDeploy) {
+        Preconditions.checkState(!fragment.getInstances().isEmpty());
+
+        // This is a load process, and it is the first fragment.
+        // we should add all BackendExecState of this fragment to needCheckBackendExecStates,
+        // so that we can check these backends' state when joining this Coordinator
+        boolean needCheckExecutionState = jobSpec.isLoadType() && fragment.getFragmentIndex() == 0;
+        boolean isEnablePipeline = jobSpec.isEnablePipeline();
+        // if pipeline is enable and current fragment contain olap table sink, in fe we will
+        // calculate the number of all tablet sinks in advance and assign them to each fragment instance
+        boolean enablePipelineTableSinkDop = isEnablePipeline && fragment.getPlanFragment().hasTableSink();
+
+        List<List<FragmentInstance>> twoStageInstancesToDeploy = ImmutableList.of(new ArrayList<>(), new ArrayList<>());
```

Comment:
what does two stage instance mean?

Reply:
This is explained in the comments in `deployFragments`:
```java
    public void deployFragments(List<ExecutionFragment> concurrentFragments) throws RpcException, UserException {
        // Divide requests of fragments in the current group to two stages.
        // - stage 1, the first request to a host, which need send descTable.
        // - stage 2, the non-first requests to a host, which needn't send descTable.
```